Documentation/bpf.rst

```diff
@@ -3354,7 +3354,355 @@ the toolchain section under the respective tools.
+callers for immediate feedback, whereas ``TC_ACT_STOLEN`` will release
+the ``skb`` through ``consume_skb()`` and pretend to upper layers that
+the transmission was successful through ``NET_XMIT_SUCCESS``. The perf's
+drop monitor which records traces of ``kfree_skb()`` will therefore
+also not see any drop indications from ``TC_ACT_STOLEN`` since its
+semantics are such that the ``skb`` has been "consumed" or queued but
+certainly not "dropped".
+
+Last but not least the ``TC_ACT_REDIRECT`` action which is available for
+tc BPF programs as well. This allows to redirect the ``skb`` to the same
+or another's device ingress or egress path together with the ``bpf_redirect()``
+helper. Being able to inject the packet into another device's ingress or
+egress direction allows for full flexibility in packet forwarding with
+BPF. There are no requirements on the target networking device other than
+being a networking device itself, there is no need to run another instance
+of ``cls_bpf`` on the target device or other such restrictions.
+
+**tc BPF FAQ**
+
+This section contains a few miscellaneous question and answer pairs related to
+tc BPF programs that are asked from time to time.
+
+* **Question:** What about ``act_bpf`` as a tc action module, is it still relevant?
+* **Answer:** Not really. Although ``cls_bpf`` and ``act_bpf`` share the same
+  functionality for tc BPF programs, ``cls_bpf`` is more flexible since it is a
+  proper superset of ``act_bpf``. The way tc works is that tc actions need to be
+  attached to tc classifiers. In order to achieve the same flexibility as ``cls_bpf``,
+  ``act_bpf`` would need to be attached to the ``cls_matchall`` classifier. As the
+  name says, this will match on every packet in order to pass them through for attached
+  tc action processing. For ``act_bpf``, this is will result in less efficient packet
+  processing than using ``cls_bpf`` in ``direct-action`` mode directly. If ``act_bpf``
+  is used in a setting with other classifiers than ``cls_bpf`` or ``cls_matchall``
+  then this will perform even worse due to the nature of operation of tc classifiers.
+  Meaning, if classifier A has a mismatch, then the packet is passed to classifier
+  B, reparsing the packet, etc, thus in the typical case there will be linear
+  processing where the packet would need to traverse N classifiers in the worst
+  case to find a match and execute ``act_bpf`` on that. Therefore, ``act_bpf`` has
+  never been largely relevant. Additionally, ``act_bpf`` does not provide a tc
+  offloading interface either compared to ``cls_bpf``.
+
+..
+
+* **Question:** Is it recommended to use ``cls_bpf`` not in ``direct-action`` mode?
+* **Answer:** No. The answer is similar to the one above in that this is otherwise
+  unable to scale for more complex processing. tc BPF can already do everything needed
+  by itself in an efficient manner and thus there is no need for anything other than
+  ``direct-action`` mode.
```

Comment:
Subquestion: What if you hit BPF instruction limits? Could you work around those by chaining programs via cls actions? (Might not be common and may/may not be something you actually want to include in the FAQ, I'm mostly just curious for myself).

Reply:
Yes, this would work. You then need to transfer state either via skb->cb or per-cpu map with single slot.